framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,527 @@
+	pthread_mutex_destroy(&g_mutex_playing_q);
+	sem_destroy(&g_sem_playing);
+
+	// Shutdown record thread
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	g_recording_live = false;
+	sem_post(&g_sem_recording);
+	pthread_join(g_pth_recording, NULL);
+	pthread_mutex_destroy(&g_mutex_recording_q);
+	sem_destroy(&g_sem_recording);
+#endif
+}
+
+media_result_t media_play(media_t *m, bool loop)
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+media_result_t media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+media_result_t media_stop(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING || m->state == MEDIA_STATE_RECORDING) {
+		m->state = MEDIA_STATE_STOPPING;
+		return MEDIA_OK;
+	}  
+	return MEDIA_ERROR;
+}
+
+media_result_t media_pause(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_result_t media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_state_t media_get_state(media_t *m)
+{
+	return m->state;
+}
+
+static media_result_t read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_type_t get_media_type(char *path, media_t * m)
+{
+	char *dot = strrchr(path, '.');
+	if (dot != NULL) {
+		if (strcmp(dot, ".wav") == 0) {
+			return MEDIA_TYPE_WAV;
+		}
+	}
+	return MEDIA_TYPE_PCM;
+}
+
+media_t *media_open(char *path, media_op_t op, media_play_type_t play_type, struct pcm_config *config)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->play_type = play_type;
+	m->config = config;
+	if (play_type == MEDIA_PLAY_TYPE_FILE) {
+		m->type = get_media_type(path, m);
+		if (m->type == MEDIA_TYPE_WAV || m->type == MEDIA_TYPE_PCM) {
+			if (op == MEDIA_OP_PLAYBACK) {
+				m->fd = open(path, O_RDONLY);
+				if (m->fd < 0) {
+					goto error_out;
+				}
+
+				if (m->type == MEDIA_TYPE_WAV) {
+					read_wav_header(m);
+				}
+				m->state = MEDIA_STATE_PLAYING;
+				return m;
+			} else if (op == MEDIA_OP_RECORD) {
+				m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
```

Comment:
As I reviewed before, fd is better than path.
Flag should not be set by framework forcely.